cluster/src/main/java/org/apache/iotdb/cluster/metadata/MetaPuller.java

```diff
@@ -0,0 +1,182 @@
+import org.apache.iotdb.db.service.IoTDB;
+import org.apache.iotdb.tsfile.write.schema.MeasurementSchema;
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class MetaPuller {
+  private static final Logger logger = LoggerFactory.getLogger(MetaPuller.class);
+  private MetaGroupMember metaGroupMember;
+
+  private MetaPuller() {
+  }
+
+  private static class MetaPullerHolder {
+
+    private static final MetaPuller INSTANCE = new MetaPuller();
+  }
+
+  public void init(MetaGroupMember metaGroupMember) {
+    this.metaGroupMember = metaGroupMember;
+  }
+
+  /**
+   * we should not use this function in other place, but only in IoTDB class
+   * @return
+   */
+  public static MetaPuller getInstance() {
+    return MetaPullerHolder.INSTANCE;
+  }
+
+  /**
+   * Pull the all timeseries schemas of given prefixPaths from remote nodes. All prefixPaths must
+   * contain the storage group.
+   */
+  public List<MeasurementSchema> pullTimeSeriesSchemas(List<String> prefixPaths)
+    throws MetadataException {
+    logger.debug("{}: Pulling timeseries schemas of {}", metaGroupMember.getName(), prefixPaths);
+    // split the paths by the data groups that will hold them
+    Map<PartitionGroup, List<String>> partitionGroupPathMap = new HashMap<>();
+    for (String prefixPath : prefixPaths) {
+      PartitionGroup partitionGroup;
+      try {
+        partitionGroup = metaGroupMember.getPartitionTable().partitionByPathTime(prefixPath, 0);
+      } catch (StorageGroupNotSetException e) {
+        // the storage group is not found locally, but may be found in the leader, retry after
+        // synchronizing with the leader
+
+        try {
+          metaGroupMember.syncLeaderWithConsistencyCheck();
+        } catch (CheckConsistencyException checkConsistencyException) {
+          throw new MetadataException(checkConsistencyException.getMessage());
+        }
+        partitionGroup = metaGroupMember.getPartitionTable().partitionByPathTime(prefixPath, 0);
+
+      }
+      partitionGroupPathMap.computeIfAbsent(partitionGroup, g -> new ArrayList<>()).add(prefixPath);
+    }
+
+    List<MeasurementSchema> schemas = new ArrayList<>();
+    // pull timeseries schema from every group involved
+    if (logger.isDebugEnabled()) {
+      logger.debug("{}: pulling schemas of {} and other {} paths from {} groups", metaGroupMember.getName(),
+        prefixPaths.get(0), prefixPaths.size() - 1,
+        partitionGroupPathMap.size());
+    }
+    for (Map.Entry<PartitionGroup, List<String>> partitionGroupListEntry : partitionGroupPathMap
+      .entrySet()) {
+      PartitionGroup partitionGroup = partitionGroupListEntry.getKey();
+      List<String> paths = partitionGroupListEntry.getValue();
+      pullTimeSeriesSchemas(partitionGroup, paths, schemas);
+    }
```

Comment:
How about using parallel pulling?

Reply:
It will be complicated, I will do it later